src/main/java/seedu/address/model/tag/TagTreeImpl.java

```diff
@@ -0,0 +1,216 @@
+        tagSubTagMap = new HashMap<>();
+        tagSuperTagMap = new HashMap<>();
+    }
+
+    /**
+     * Copys the argument {@code toCopy} tree into this tree.
+     */
+    public TagTreeImpl(ReadOnlyTagTree toCopy) {
+        super();
+        copy(toCopy);
+    }
+
+    /**
+     * Creates a TagTreeImpl with the given {@code tagSubTagMap} and {@code tagSuperTagMap}.
+     */
+    public TagTreeImpl(Map<Tag, Set<Tag>> tagSubTagMap, Map<Tag, Set<Tag>> tagSuperTagMap) {
+        this.tagSubTagMap = tagSubTagMap;
+        this.tagSuperTagMap = tagSuperTagMap;
+    }
+
+    @Override
+    public void copy(ReadOnlyTagTree toCopy) {
+        tagSubTagMap = new HashMap<>();
+        tagSuperTagMap = new HashMap<>();
+
+        toCopy.getTagSubTagMap().entrySet().stream().forEach(entry -> {
+            // Adds a copy of the key-value mapping to the subtag map.
+            tagSubTagMap.put(entry.getKey(), new HashSet<>(Set.copyOf(entry.getValue())));
+
+            // For each sub-tag, adds an equivalent reverse mapping to the super tag map.
+            entry.getValue().stream().forEach(subTag -> tagSuperTagMap.merge(
+                subTag, new HashSet<>(Set.of(entry.getKey())), (set1, set2) -> {
+                    set1.addAll(set2);
+                    return set1; })); });
+    }
+
+    @Override
+    public Set<Tag> getSuperTags() {
+        return tagSubTagMap.keySet();
+    }
+
+    @Override
+    public Map<Tag, Set<Tag>> getTagSubTagMap() {
+        Map<Tag, Set<Tag>> tempMap = new HashMap<>();
+        tagSubTagMap.entrySet().stream().forEach(entry ->
+                tempMap.put(entry.getKey(), Set.copyOf(entry.getValue())));
+        return Map.copyOf(tempMap);
+    }
+
+    Map<Tag, Set<Tag>> getInternalTagSubTagMap() {
+        return tagSubTagMap;
+    }
+
+    @Override
+    public Set<Tag> getSubTagsOf(Tag tag) {
+        return tagSubTagMap.containsKey(tag) ? Set.copyOf(tagSubTagMap.get(tag)) : Set.of();
+    }
+
+    @Override
+    public Set<Tag> getSubTagsRecursive(Tag tag) {
+        Set<Tag> finalSet = new HashSet<>();
+        Consumer<Tag> consumer = new Consumer<Tag>() {
+            @Override
+            public void accept(Tag tag) {
+                if (tagSubTagMap.get(tag) == null) {
+                    return;
+                }
+                finalSet.addAll(tagSubTagMap.get(tag));
+                tagSubTagMap.get(tag).forEach(subtag -> accept(subtag));
+            }
+        };
+        consumer.accept(tag);
+        return finalSet;
+    }
+
+    @Override
+    public void addSubTagTo(Tag superTag, Tag subTag) {
+        assert !superTag.equals(Tag.ALL_TAGS_TAG) && !subTag.equals(Tag.ALL_TAGS_TAG);
+
+        if (isSubTagOf(subTag, superTag)) {
+            throw new IllegalArgumentException(String.format(MESSAGE_CYCLIC_RELATIONSHIP, superTag, subTag));
+        }
+        addToMapSet(tagSubTagMap, superTag, subTag);
+        addToMapSet(tagSuperTagMap, subTag, superTag);
+    }
```

Comment:
should these have documentation?

Reply:
The documentation of its purpose is in the `TagTree` abstract class. I wanted to leave open the implementing class in the event someone could come up with a better underlying system for it. Admittedly though, many classes do depend on associate with `TagTree` and as such, depend on `TagTreeImpl` anyway.

Reply:
makes sense.